mqtt/mqtt-broker/src/state_change.rs

```diff
@@ -0,0 +1,413 @@
+            session
+                .subscriptions()
+                .map(|s| s.keys().map(String::as_str).collect())
+        });
+
+        Self::Subscriptions(client_id, subscriptions)
+    }
+
+    pub fn new_connection_change(sessions: &'a HashMap<ClientId, Session>) -> Self {
+        let connections = sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Connections(connections)
+    }
+
+    pub fn new_session_change(sessions: &'a HashMap<ClientId, Session>) -> Self {
+        let sessions = sessions
+            .iter()
+            .filter_map(|(client_id, session)| match session {
+                Session::Transient(_) => Some(client_id),
+                Session::Persistent(_) => Some(client_id),
+                Session::Offline(_) => Some(client_id),
+                _ => None,
+            })
+            .collect();
+
+        Self::Sessions(sessions)
+    }
+}
+
+impl<'a> TryFrom<StateChange<'a>> for proto::Publication {
+    type Error = Error;
+
+    fn try_from(state: StateChange<'a>) -> Result<Self, Error> {
+        Ok(match state {
+            StateChange::Subscriptions(client_id, subscriptions) => {
+                let payload = subscriptions
+                    .map(|subscriptions| serde_json::to_string(&subscriptions))
+                    .transpose()?
+                    .map_or_else(bytes::Bytes::new, |json| json.into());
+
+                proto::Publication {
+                    topic_name: format!("$edgehub/subscriptions/{}", client_id),
+                    qos: STATE_CHANGE_QOS,
+                    retain: true,
+                    payload,
+                }
+            }
+            StateChange::Connections(connections) => proto::Publication {
+                topic_name: "$edgehub/connected".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&connections)?.into(),
+            },
+            StateChange::Sessions(sessions) => proto::Publication {
+                topic_name: "$edgehub/sessions".to_owned(),
+                qos: STATE_CHANGE_QOS,
+                retain: true,
+                payload: serde_json::to_string(&sessions)?.into(),
+            },
+        })
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use std::collections::{HashMap, VecDeque};
+    use std::convert::TryInto;
+    use std::str::FromStr;
+
+    use super::{StateChange, STATE_CHANGE_QOS};
+    use crate::broker::tests::{connection_handle, is_notify_equal, persistent_connect};
+    use crate::session::{Session, SessionState};
+    use crate::subscription::{Subscription, TopicFilter};
+    use crate::{AuthId, ClientId, ConnReq};
+    use mqtt3::proto;
```

Comment:
missed new line

Reply:
cargo fmt removes it for some reason

Reply:
weird, it removes it when there's 1 newline, but leaves a new line when I add 2...